File: kafka-streams-scala/7-windowed-leaderboard-scala/src/main/scala/com/github/simplesteph/udemy/kafka/streams/VictoryTimestampKeyExtractor.scala
package com.github.simplesteph.udemy.kafka.streams

import com.github.simplesteph.Victory
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.streams.processor.TimestampExtractor

import scala.util.Try


class VictoryTimestampKeyExtractor extends TimestampExtractor {

  override def extract(record: ConsumerRecord[AnyRef, AnyRef], previousTimestamp: Long): Long =
    Try(record.value.asInstanceOf[Victory].timestamp.toEpochMilli).getOrElse(previousTimestamp)

}

File: kafka-streams-java/7-windowed-leaderboard/src/main/java/com/github/simplesteph/udemy/kafka/streams/VictoryTimestampKeyExtractor.java
package com.github.simplesteph.udemy.kafka.streams;

import com.github.simplesteph.Victory;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.streams.processor.TimestampExtractor;

public class VictoryTimestampKeyExtractor implements TimestampExtractor {

    @Override
    public long extract(ConsumerRecord<Object, Object> record, long previousTimestamp) {
        long newTimestamp;
        try {
            newTimestamp = ((Victory) record.value()).getTimestamp().toEpochMilli();
        } catch (Throwable e) {
            newTimestamp = previousTimestamp;
        }
        return newTimestamp;
    }
}
